src/librustc/middle/resolve_lifetime.rs

```diff
@@ -299,22 +316,92 @@ type ScopeRef<'a> = &'a Scope<'a>;
 
 const ROOT_SCOPE: ScopeRef<'static> = &Scope::Root;
 
-pub fn krate(
-    sess: &Session,
-    cstore: &CrateStore,
-    hir_map: &Map,
-) -> Result<NamedRegionMap, ErrorReported> {
-    let krate = hir_map.krate();
+pub fn provide(providers: &mut ty::maps::Providers) {
+    *providers = ty::maps::Providers {
+        resolve_lifetimes,
+
+        named_region_map: |tcx, id| {
+            let id = LocalDefId::from_def_id(DefId::local(id)); // (*)
+            tcx.resolve_lifetimes(LOCAL_CRATE).defs.get(&id).cloned()
+        },
+
+        is_late_bound_map: |tcx, id| {
+            let id = LocalDefId::from_def_id(DefId::local(id)); // (*)
+            tcx.resolve_lifetimes(LOCAL_CRATE)
+                .late_bound
+                .get(&id)
+                .cloned()
+        },
+
+        object_lifetime_defaults_map: |tcx, id| {
+            let id = LocalDefId::from_def_id(DefId::local(id)); // (*)
+            tcx.resolve_lifetimes(LOCAL_CRATE)
+                .object_lifetime_defaults
+                .get(&id)
+                .cloned()
+        },
+
+        ..*providers
+    };
+
+    // (*) FIXME the query should be defined to take a LocalDefId
+}
+
+/// Computes the `ResolveLifetimes` map that contains data for the
+/// entire crate. You should not read the result of this query
+/// directly, but rather use `named_region_map`, `is_late_bound_map`,
+/// etc.
+fn resolve_lifetimes<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    for_krate: CrateNum,
+) -> Rc<ResolveLifetimes> {
+    assert_eq!(for_krate, LOCAL_CRATE);
+
+    let named_region_map = krate(tcx).unwrap_or_default();
+
+    let mut defs = FxHashMap();
+    for (k, v) in named_region_map.defs {
+        let hir_id = tcx.hir.node_to_hir_id(k);
+        let map = defs.entry(hir_id.owner_local_def_id())
+            .or_insert_with(|| Rc::new(FxHashMap()));
+        Rc::get_mut(map).unwrap().insert(hir_id.local_id, v);
+    }
+    let mut late_bound = FxHashMap();
+    for k in named_region_map.late_bound {
+        let hir_id = tcx.hir.node_to_hir_id(k);
+        let map = late_bound
+            .entry(hir_id.owner_local_def_id())
+            .or_insert_with(|| Rc::new(FxHashSet()));
+        Rc::get_mut(map).unwrap().insert(hir_id.local_id);
+    }
+    let mut object_lifetime_defaults = FxHashMap();
+    for (k, v) in named_region_map.object_lifetime_defaults {
+        let hir_id = tcx.hir.node_to_hir_id(k);
+        let map = object_lifetime_defaults
+            .entry(hir_id.owner_local_def_id())
+            .or_insert_with(|| Rc::new(FxHashMap()));
+        Rc::get_mut(map)
+            .unwrap()
+            .insert(hir_id.local_id, Rc::new(v));
+    }
+
+    Rc::new(ResolveLifetimes {
+        defs,
+        late_bound,
+        object_lifetime_defaults,
+    })
+}
+
+fn krate<'tcx>(tcx: TyCtxt<'_, 'tcx, 'tcx>) -> Result<NamedRegionMap, ErrorReported> {
```

Comment:
Can this function actually return `ErrorReported` now?

Reply:
You're right, I should fix that. A holdover from older code.